<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Debug for the ErrorKind - chainerror</title>
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="A tutorial for the chainerror rust crate.">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" rel="stylesheet" type="text/css">
        <link href="https://fonts.googleapis.com/css?family=Source+Code+Pro:500" rel="stylesheet" type="text/css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        

        
    </head>
    <body class="light">
        <!-- Provide site root to javascript -->
        <script type="text/javascript">
            var path_to_root = "";
            var default_theme = "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script type="text/javascript">
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { } 
            if (theme === null || theme === undefined) { theme = default_theme; }
            document.body.className = theme;
            document.querySelector('html').className = theme + ' js';
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var html = document.querySelector('html');
            var sidebar = 'hidden';
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <ol class="chapter"><li class="affix"><a href="index.html">chainerror</a></li><li><a href="tutorial1.html"><strong aria-hidden="true">1.</strong> Simple String Errors</a></li><li><a href="tutorial2.html"><strong aria-hidden="true">2.</strong> Simple Chained String Errors</a></li><li><a href="tutorial3.html"><strong aria-hidden="true">3.</strong> Mapping Errors</a></li><li><a href="tutorial4.html"><strong aria-hidden="true">4.</strong> Saving coding chars</a></li><li><a href="tutorial5.html"><strong aria-hidden="true">5.</strong> The source() of Errors</a></li><li><a href="tutorial6.html"><strong aria-hidden="true">6.</strong> Downcast the Errors</a></li><li><a href="tutorial7.html"><strong aria-hidden="true">7.</strong> The root cause of all Errors</a></li><li><a href="tutorial8.html"><strong aria-hidden="true">8.</strong> Finding an Error cause</a></li><li><a href="tutorial9.html"><strong aria-hidden="true">9.</strong> Selective Error Handling</a></li><li><a href="tutorial10.html"><strong aria-hidden="true">10.</strong> ErrorKind to the rescue</a></li><li><a href="tutorial11.html" class="active"><strong aria-hidden="true">11.</strong> Debug for the ErrorKind</a></li><li class="affix"><a href="end.html">The End</a></li></ol>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                
                <div id="menu-bar" class="menu-bar">
                    <div id="menu-bar-sticky-container">
                        <div class="left-buttons">
                            <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                                <i class="fa fa-bars"></i>
                            </button>
                            <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                                <i class="fa fa-paint-brush"></i>
                            </button>
                            <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                                <li role="none"><button role="menuitem" class="theme" id="light">Light (default)</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                                <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                            </ul>
                            
                            <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                                <i class="fa fa-search"></i>
                            </button>
                            
                        </div>

                        <h1 class="menu-title">chainerror</h1> 

                        <div class="right-buttons">
                            <a href="print.html" title="Print this book" aria-label="Print this book">
                                <i id="print-button" class="fa fa-print"></i>
                            </a>
                            
                        </div>
                    </div>
                </div>

                
                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" name="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>
                

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script type="text/javascript">
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <a class="header" href="#debug-for-the-errorkind" id="debug-for-the-errorkind"><h1>Debug for the ErrorKind</h1></a>
<p>One small improvement at the end of the tutorial is to fix the debug output of
<code>Func1ErrorKind</code>. As you probably noticed, the output doesn't say much of the enum.</p>
<pre><code>Debug Error:
src/main.rs:35: Func2
[…]
</code></pre>
<p>As a lazy shortcut, we implement <code>Debug</code> by calling <code>Display</code> and end up with</p>
<pre><code>Debug Error:
src/main.rs:40: func1 error calling func2
[…}
</code></pre>
<p>which gives us a lot more detail.</p>
<p>To create your own Errors, you might find <a href="https://crates.io">crates</a> which create enum <code>Display+Debug</code> via derive macros.</p>
<p>Also noteworthy is <a href="https://crates.io/crates/custom_error">custom_error</a> to define your custom errors,
which can then be used with <code>chainerror</code>.</p>
<pre><pre class="playpen"><code class="language-rust">use crate::chainerror::*;
use std::error::Error;
use std::io;
use std::result::Result;

fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
    Err(io::Error::from(io::ErrorKind::NotFound))?;
    Ok(())
}

derive_str_cherr!(Func2Error);

fn func2() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
    let filename = &quot;foo.txt&quot;;
    do_some_io().map_err(mstrerr!(Func2Error, &quot;Error reading '{}'&quot;, filename))?;
    Ok(())
}

enum Func1ErrorKind {
    Func2,
    IO(String),
}

impl ::std::fmt::Display for Func1ErrorKind {
    fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
        match self {
            Func1ErrorKind::Func2 =&gt; write!(f, &quot;func1 error calling func2&quot;),
            Func1ErrorKind::IO(filename) =&gt; write!(f, &quot;Error reading '{}'&quot;, filename),
        }
    }
}

impl ::std::fmt::Debug for Func1ErrorKind {
    fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
        write!(f, &quot;{}&quot;, self)
    }
}

fn func1() -&gt; ChainResult&lt;(), Func1ErrorKind&gt; {
    func2().map_err(|e| cherr!(e, Func1ErrorKind::Func2))?;
    let filename = String::from(&quot;bar.txt&quot;);
    do_some_io().map_err(|e| cherr!(e, Func1ErrorKind::IO(filename)))?;
    Ok(())
}

fn main() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
    if let Err(e) = func1() {
        match e.kind() {
            Func1ErrorKind::Func2 =&gt; eprintln!(&quot;Main Error Report: func1 error calling func2&quot;),
            Func1ErrorKind::IO(filename) =&gt; {
                eprintln!(&quot;Main Error Report: func1 error reading '{}'&quot;, filename)
            }
        }

        if let Some(e) = e.find_chain_cause::&lt;Func2Error&gt;() {
            eprintln!(&quot;\nError reported by Func2Error: {}&quot;, e)
        }

        eprintln!(&quot;\nDebug Error:\n{:?}&quot;, e);
    }
    Ok(())
}
# #[allow(dead_code)]
# mod chainerror {
# /*!
# 
# `chainerror` provides an error backtrace without doing a real backtrace, so even after you `strip` your
# binaries, you still have the error backtrace.
# 
# `chainerror` has no dependencies!
# 
# `chainerror` uses `.source()` of `std::error::Error` along with `line()!` and `file()!` to provide a nice debug error backtrace.
# It encapsulates all types, which have `Display + Debug` and can store the error cause internally.
# 
# Along with the `ChainError&lt;T&gt;` struct, `chainerror` comes with some useful helper macros to save a lot of typing.
# 
# ## Features
# 
# `no-fileline`
# : completely turn off storing filename and line
# 
# `display-cause`
# : turn on printing a backtrace of the errors in `Display`
# 
# `no-debug-cause`
# : turn off printing a backtrace of the errors in `Debug`
# 
# 
# # Tutorial
# 
# Read the [Tutorial](https://haraldh.github.io/chainerror/tutorial1.html)
# 
# # Examples
# 
# ~~~rust
# use chainerror::*;
# use std::error::Error;
# use std::io;
# use std::result::Result;
# 
# fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     Err(io::Error::from(io::ErrorKind::NotFound))?;
#     Ok(())
# }
# 
# fn func2() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     let filename = &quot;foo.txt&quot;;
#     do_some_io().map_err(mstrerr!(&quot;Error reading '{}'&quot;, filename))?;
#     Ok(())
# }
# 
# fn func1() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     func2().map_err(mstrerr!(&quot;func1 error&quot;))?;
#     Ok(())
# }
# 
# fn main() {
#     if let Err(e) = func1() {
#         assert_eq!(
#             format!(&quot;\n{:?}\n&quot;, e), r#&quot;
# src/lib.rs:20: func1 error
# Caused by:
# src/lib.rs:15: Error reading 'foo.txt'
# Caused by:
# Kind(NotFound)
# &quot;#
#         );
#     }
# #    else {
# #        unreachable!();
# #    }
# }
# ~~~
# 
# 
# ~~~rust
# use chainerror::*;
# use std::error::Error;
# use std::io;
# use std::result::Result;
# 
# fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     Err(io::Error::from(io::ErrorKind::NotFound))?;
#     Ok(())
# }
# 
# fn func3() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     let filename = &quot;foo.txt&quot;;
#     do_some_io().map_err(mstrerr!(&quot;Error reading '{}'&quot;, filename))?;
#     Ok(())
# }
# 
# derive_str_cherr!(Func2Error);
# 
# fn func2() -&gt; ChainResult&lt;(), Func2Error&gt; {
#     func3().map_err(mstrerr!(Func2Error, &quot;func2 error: calling func3&quot;))?;
#     Ok(())
# }
# 
# enum Func1Error {
#     Func2,
#     IO(String),
# }
# 
# impl ::std::fmt::Display for Func1Error {
#     fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
#         match self {
#             Func1Error::Func2 =&gt; write!(f, &quot;func1 error calling func2&quot;),
#             Func1Error::IO(filename) =&gt; write!(f, &quot;Error reading '{}'&quot;, filename),
#         }
#     }
# }
# 
# impl ::std::fmt::Debug for Func1Error {
#     fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
#         write!(f, &quot;{}&quot;, self)
#     }
# }
# 
# fn func1() -&gt; ChainResult&lt;(), Func1Error&gt; {
#     func2().map_err(|e| cherr!(e, Func1Error::Func2))?;
#     let filename = String::from(&quot;bar.txt&quot;);
#     do_some_io().map_err(|e| cherr!(e, Func1Error::IO(filename)))?;
#     Ok(())
# }
# 
# fn main() {
#     if let Err(e) = func1() {
#         assert!(
#             match e.kind() {
#                 Func1Error::Func2 =&gt; {
#                     eprintln!(&quot;Main Error Report: func1 error calling func2&quot;);
#                     true
#                 }
#                 Func1Error::IO(filename) =&gt; {
#                     eprintln!(&quot;Main Error Report: func1 error reading '{}'&quot;, filename);
#                     false
#                 }
#             }
#         );
# 
#         assert!(e.find_chain_cause::&lt;Func2Error&gt;().is_some());
# 
#         if let Some(e) = e.find_chain_cause::&lt;Func2Error&gt;() {
#             eprintln!(&quot;\nError reported by Func2Error: {}&quot;, e)
#         }
# 
# 
#         assert!(e.root_cause().is_some());
# 
#         if let Some(e) = e.root_cause() {
#             let ioerror = e.downcast_ref::&lt;io::Error&gt;().unwrap();
#             eprintln!(&quot;\nThe root cause was: std::io::Error: {:#?}&quot;, ioerror);
#         }
# 
#         assert_eq!(
#             format!(&quot;\n{:?}\n&quot;, e), r#&quot;
# src/lib.rs:47: func1 error calling func2
# Caused by:
# src/lib.rs:22: Func2Error(func2 error: calling func3)
# Caused by:
# src/lib.rs:15: Error reading 'foo.txt'
# Caused by:
# Kind(NotFound)
# &quot;#
#         );
#     }
# #    else {
# #        unreachable!();
# #    }
# }
# ~~~
# 
# !*/
# 
# use std::any::TypeId;
# use std::error::Error;
# use std::fmt::{Debug, Display, Formatter, Result};
# 
# /** chains an inner error kind `T` with a causing error
# **/
# pub struct ChainError&lt;T&gt; {
#     #[cfg(not(feature = &quot;no-fileline&quot;))]
#     occurrence: Option&lt;(u32, &amp;'static str)&gt;,
#     kind: T,
#     error_cause: Option&lt;Box&lt;dyn Error + 'static&gt;&gt;,
# }
# 
# /// convenience type alias
# pub type ChainResult&lt;O, E&gt; = std::result::Result&lt;O, ChainError&lt;E&gt;&gt;;
# 
# impl&lt;T: 'static + Display + Debug&gt; ChainError&lt;T&gt; {
#     #[cfg(not(feature = &quot;no-fileline&quot;))]
#     /// Use the `cherr!()` or `mstrerr!()` macro instead of calling this directly
#     pub fn new(
#         kind: T,
#         error_cause: Option&lt;Box&lt;dyn Error + 'static&gt;&gt;,
#         occurrence: Option&lt;(u32, &amp;'static str)&gt;,
#     ) -&gt; Self {
#         Self {
#             occurrence,
#             kind,
#             error_cause,
#         }
#     }
# 
#     #[cfg(feature = &quot;no-fileline&quot;)]
#     /// Use the `cherr!()` or `mstrerr!()` macro instead of calling this directly
#     pub fn new(
#         kind: T,
#         error_cause: Option&lt;Box&lt;dyn Error + 'static&gt;&gt;,
#         _occurrence: Option&lt;(u32, &amp;'static str)&gt;,
#     ) -&gt; Self {
#         Self { kind, error_cause }
#     }
# 
#     /// return the root cause of the error chain, if any exists
#     pub fn root_cause(&amp;self) -&gt; Option&lt;&amp;(dyn Error + 'static)&gt; {
#         let mut cause = self as &amp;(dyn Error + 'static);
#         while let Some(c) = cause.source() {
#             cause = c;
#         }
#         Some(cause)
#     }
# 
#     /** find the first error cause of type U, if any exists
# 
#     # Examples
# 
#     ~~~rust
#     # use crate::chainerror::*;
#     # use std::error::Error;
#     # use std::io;
#     # use std::result::Result;
#     #
#     fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#         Err(io::Error::from(io::ErrorKind::NotFound))?;
#         Ok(())
#     }
# 
#     derive_str_cherr!(Func2Error);
# 
#     fn func2() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#         let filename = &quot;foo.txt&quot;;
#         do_some_io().map_err(mstrerr!(Func2Error, &quot;Error reading '{}'&quot;, filename))?;
#         Ok(())
#     }
# 
#     derive_str_cherr!(Func1Error);
# 
#     fn func1() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#         func2().map_err(mstrerr!(Func1Error, &quot;func1 error&quot;))?;
#         Ok(())
#     }
# 
#     fn main() {
#         if let Err(e) = func1() {
#             if let Some(f1err) = e.downcast_chain_ref::&lt;Func1Error&gt;() {
# 
#                 assert!(f1err.find_cause::&lt;io::Error&gt;().is_some());
# 
#                 assert!(f1err.find_chain_cause::&lt;Func2Error&gt;().is_some());
#             }
#     #        else {
#     #            panic!();
#     #        }
#         }
#     #    else {
#     #         unreachable!();
#     #    }
#     }
#     ~~~
#     **/
#     pub fn find_cause&lt;U: Error + 'static&gt;(&amp;self) -&gt; Option&lt;&amp;U&gt; {
#         let mut cause = self as &amp;(dyn Error + 'static);
#         loop {
#             if cause.is::&lt;U&gt;() {
#                 return cause.downcast_ref::&lt;U&gt;();
#             }
# 
#             match cause.source() {
#                 Some(c) =&gt; cause = c,
#                 None =&gt; return None,
#             }
#         }
#     }
# 
#     /** find the first error cause of type ChainError&lt;U&gt;, if any exists
# 
#     Same as `find_cause`, but hides the `ChainError&lt;U&gt;` implementation internals
# 
#     # Examples
# 
#     ~~~rust,ignore
#     /// Instead of writing
#     err.find_cause::&lt;ChainError&lt;FooError&gt;&gt;();
# 
#     /// leave out the ChainError&lt;T&gt; implementation detail
#     err.find_chain_cause::&lt;FooError&gt;();
#     ~~~
# 
#     **/
#     pub fn find_chain_cause&lt;U: Error + 'static&gt;(&amp;self) -&gt; Option&lt;&amp;ChainError&lt;U&gt;&gt; {
#         let mut cause = self as &amp;(dyn Error + 'static);
#         loop {
#             if cause.is::&lt;ChainError&lt;U&gt;&gt;() {
#                 return cause.downcast_ref::&lt;ChainError&lt;U&gt;&gt;();
#             }
# 
#             match cause.source() {
#                 Some(c) =&gt; cause = c,
#                 None =&gt; return None,
#             }
#         }
#     }
# 
#     /** return a reference to T of `ChainError&lt;T&gt;`
# 
#     # Examples
# 
#     ~~~rust
#     # use crate::chainerror::*;
#     # use std::error::Error;
#     # use std::io;
#     # use std::result::Result;
#     #
#     fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#         Err(io::Error::from(io::ErrorKind::NotFound))?;
#         Ok(())
#     }
# 
#     derive_str_cherr!(Func2Error);
# 
#     fn func2() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#         let filename = &quot;foo.txt&quot;;
#         do_some_io().map_err(mstrerr!(Func2Error, &quot;Error reading '{}'&quot;, filename))?;
#         Ok(())
#     }
# 
#     #[derive(Debug)]
#     enum Func1ErrorKind {
#         Func2,
#         IO(String),
#     }
# 
#     // impl ::std::fmt::Display for Func1ErrorKind {…}
#     # impl ::std::fmt::Display for Func1ErrorKind {
#     #     fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
#     #         match self {
#     #             Func1ErrorKind::Func2 =&gt; write!(f, &quot;func1 error calling func2&quot;),
#     #             Func1ErrorKind::IO(filename) =&gt; write!(f, &quot;Error reading '{}'&quot;, filename),
#     #         }
#     #     }
#     # }
# 
#     fn func1() -&gt; ChainResult&lt;(), Func1ErrorKind&gt; {
#         func2().map_err(|e| cherr!(e, Func1ErrorKind::Func2))?;
#         do_some_io().map_err(|e| cherr!(e, Func1ErrorKind::IO(&quot;bar.txt&quot;.into())))?;
#         Ok(())
#     }
# 
#     fn main() {
#         if let Err(e) = func1() {
#             match e.kind() {
#                 Func1ErrorKind::Func2 =&gt; {},
#                 Func1ErrorKind::IO(filename) =&gt; panic!(),
#             }
#         }
#     #    else {
#     #         unreachable!();
#     #    }
#     }
#     ~~~
# 
#     **/
#     pub fn kind&lt;'a&gt;(&amp;'a self) -&gt; &amp;'a T {
#         &amp;self.kind
#     }
# }
# 
# /** convenience trait to hide the `ChainError&lt;T&gt;` implementation internals
# **/
# pub trait ChainErrorDown {
#     /** test if of type `ChainError&lt;T&gt;`
#      **/
#     fn is_chain&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; bool;
#     /** downcast to a reference of `ChainError&lt;T&gt;`
#      **/
#     fn downcast_chain_ref&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; Option&lt;&amp;ChainError&lt;T&gt;&gt;;
#     /** downcast to a mutable reference of `ChainError&lt;T&gt;`
#      **/
#     fn downcast_chain_mut&lt;T: 'static + Display + Debug&gt;(&amp;mut self) -&gt; Option&lt;&amp;mut ChainError&lt;T&gt;&gt;;
# }
# 
# impl&lt;U: 'static + Display + Debug&gt; ChainErrorDown for ChainError&lt;U&gt; {
#     fn is_chain&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; bool {
#         TypeId::of::&lt;T&gt;() == TypeId::of::&lt;U&gt;()
#     }
# 
#     fn downcast_chain_ref&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; Option&lt;&amp;ChainError&lt;T&gt;&gt; {
#         if self.is_chain::&lt;T&gt;() {
#             unsafe { Some(&amp;*(self as *const dyn Error as *const &amp;ChainError&lt;T&gt;)) }
#         } else {
#             None
#         }
#     }
# 
#     fn downcast_chain_mut&lt;T: 'static + Display + Debug&gt;(&amp;mut self) -&gt; Option&lt;&amp;mut ChainError&lt;T&gt;&gt; {
#         if self.is_chain::&lt;T&gt;() {
#             unsafe { Some(&amp;mut *(self as *mut dyn Error as *mut &amp;mut ChainError&lt;T&gt;)) }
#         } else {
#             None
#         }
#     }
# }
# 
# impl ChainErrorDown for dyn Error + 'static {
#     fn is_chain&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; bool {
#         self.is::&lt;ChainError&lt;T&gt;&gt;()
#     }
# 
#     fn downcast_chain_ref&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; Option&lt;&amp;ChainError&lt;T&gt;&gt; {
#         self.downcast_ref::&lt;ChainError&lt;T&gt;&gt;()
#     }
# 
#     fn downcast_chain_mut&lt;T: 'static + Display + Debug&gt;(&amp;mut self) -&gt; Option&lt;&amp;mut ChainError&lt;T&gt;&gt; {
#         self.downcast_mut::&lt;ChainError&lt;T&gt;&gt;()
#     }
# }
# 
# impl ChainErrorDown for dyn Error + 'static + Send {
#     fn is_chain&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; bool {
#         self.is::&lt;ChainError&lt;T&gt;&gt;()
#     }
# 
#     fn downcast_chain_ref&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; Option&lt;&amp;ChainError&lt;T&gt;&gt; {
#         self.downcast_ref::&lt;ChainError&lt;T&gt;&gt;()
#     }
# 
#     fn downcast_chain_mut&lt;T: 'static + Display + Debug&gt;(&amp;mut self) -&gt; Option&lt;&amp;mut ChainError&lt;T&gt;&gt; {
#         self.downcast_mut::&lt;ChainError&lt;T&gt;&gt;()
#     }
# }
# 
# impl ChainErrorDown for dyn Error + 'static + Send + Sync {
#     fn is_chain&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; bool {
#         self.is::&lt;ChainError&lt;T&gt;&gt;()
#     }
# 
#     fn downcast_chain_ref&lt;T: 'static + Display + Debug&gt;(&amp;self) -&gt; Option&lt;&amp;ChainError&lt;T&gt;&gt; {
#         self.downcast_ref::&lt;ChainError&lt;T&gt;&gt;()
#     }
# 
#     fn downcast_chain_mut&lt;T: 'static + Display + Debug&gt;(&amp;mut self) -&gt; Option&lt;&amp;mut ChainError&lt;T&gt;&gt; {
#         self.downcast_mut::&lt;ChainError&lt;T&gt;&gt;()
#     }
# }
# 
# impl&lt;T: 'static + Display + Debug&gt; Error for ChainError&lt;T&gt; {
#     fn source(&amp;self) -&gt; Option&lt;&amp;(dyn Error + 'static)&gt; {
#         if let Some(ref e) = self.error_cause {
#             Some(e.as_ref())
#         } else {
#             None
#         }
#     }
# }
# 
# impl&lt;T: 'static + Display + Debug&gt; Error for &amp;ChainError&lt;T&gt; {
#     fn source(&amp;self) -&gt; Option&lt;&amp;(dyn Error + 'static)&gt; {
#         if let Some(ref e) = self.error_cause {
#             Some(e.as_ref())
#         } else {
#             None
#         }
#     }
# }
# 
# impl&lt;T: 'static + Display + Debug&gt; Error for &amp;mut ChainError&lt;T&gt; {
#     fn source(&amp;self) -&gt; Option&lt;&amp;(dyn Error + 'static)&gt; {
#         if let Some(ref e) = self.error_cause {
#             Some(e.as_ref())
#         } else {
#             None
#         }
#     }
# }
# 
# impl&lt;T: 'static + Display + Debug&gt; Display for ChainError&lt;T&gt; {
#     fn fmt(&amp;self, f: &amp;mut Formatter) -&gt; Result {
#         write!(f, &quot;{}&quot;, self.kind)?;
# 
#         #[cfg(feature = &quot;display-cause&quot;)]
#         {
#             if let Some(e) = self.source() {
#                 writeln!(f, &quot;\nCaused by:&quot;)?;
#                 Display::fmt(&amp;e, f)?;
#             }
#         }
#         Ok(())
#     }
# }
# 
# impl&lt;T: 'static + Display + Debug&gt; Debug for ChainError&lt;T&gt; {
#     fn fmt(&amp;self, f: &amp;mut Formatter) -&gt; Result {
#         #[cfg(not(feature = &quot;no-fileline&quot;))]
#         {
#             if let Some(o) = self.occurrence {
#                 write!(f, &quot;{}:{}: &quot;, o.1, o.0)?;
#             }
#         }
# 
#         if self.is_chain::&lt;String&gt;() {
#             Display::fmt(&amp;self.kind, f)?;
#         } else {
#             Debug::fmt(&amp;self.kind, f)?;
#         }
# 
#         #[cfg(not(feature = &quot;no-debug-cause&quot;))]
#         {
#             if let Some(e) = self.source() {
#                 writeln!(f, &quot;\nCaused by:&quot;)?;
#                 Debug::fmt(&amp;e, f)?;
#             }
#         }
#         Ok(())
#     }
# }
# 
# /** creates a new `ChainError&lt;T&gt;`
# 
# # Examples
# 
# Create a new ChainError&lt;FooError&gt;, where `FooError` must implement `Display` and `Debug`.
# ~~~rust
# # use chainerror::*;
# #
# # #[derive(Debug)]
# enum FooError {
#     Bar,
#     Baz(&amp;'static str),
# }
# #
# # impl ::std::fmt::Display for FooError {
# #     fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
# #         match self {
# #             FooError::Bar =&gt; write!(f, &quot;Bar Error&quot;),
# #             FooError::Baz(s) =&gt; write!(f, &quot;Baz Error: '{}'&quot;, s),
# #         }
# #     }
# # }
# 
# // impl ::std::fmt::Display for FooError
# 
# fn do_some_stuff() -&gt; bool {
#     false
# }
# 
# fn func() -&gt; ChainResult&lt;(), FooError&gt; {
#     if ! do_some_stuff() {
#         Err(cherr!(FooError::Baz(&quot;Error&quot;)))?;
#     }
#     Ok(())
# }
# #
# # pub fn main() {
# #     match func().unwrap_err().kind() {
# #         FooError::Baz(s) if s == &amp;&quot;Error&quot; =&gt; {},
# #         _ =&gt; panic!(),
# #     }
# # }
# ~~~
# 
# Additionally an error cause can be added.
# 
# ~~~rust
# # use chainerror::*;
# # use std::io;
# # use std::error::Error;
# #
# # #[derive(Debug)]
# # enum FooError {
# #     Bar,
# #     Baz(&amp;'static str),
# # }
# #
# # impl ::std::fmt::Display for FooError {
# #     fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
# #         match self {
# #             FooError::Bar =&gt; write!(f, &quot;Bar Error&quot;),
# #             FooError::Baz(s) =&gt; write!(f, &quot;Baz Error: '{}'&quot;, s),
# #         }
# #     }
# # }
# #
# fn do_some_stuff() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     Err(io::Error::from(io::ErrorKind::NotFound))?;
#     Ok(())
# }
# 
# fn func() -&gt; ChainResult&lt;(), FooError&gt; {
#     do_some_stuff().map_err(
#         |e| cherr!(e, FooError::Baz(&quot;Error&quot;))
#     )?;
#     Ok(())
# }
# #
# # pub fn main() {
# #     match func().unwrap_err().kind() {
# #         FooError::Baz(s) if s == &amp;&quot;Error&quot; =&gt; {},
# #         _ =&gt; panic!(),
# #     }
# # }
# ~~~
# 
# **/
# #[macro_export]
# macro_rules! cherr {
#     ( $k:expr ) =&gt; {
#         ChainError::&lt;_&gt;::new($k, None, Some((line!(), file!())))
#     };
#     ( $e:expr, $k:expr ) =&gt; {
#         ChainError::&lt;_&gt;::new($k, Some(Box::from($e)), Some((line!(), file!())))
#     };
# }
# 
# /** convenience macro for |e| cherr!(e, format!(…))
# 
# # Examples
# 
# ~~~rust
# # use crate::chainerror::*;
# # use std::error::Error;
# # use std::io;
# # use std::result::Result;
# #
# # fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
# #     Err(io::Error::from(io::ErrorKind::NotFound))?;
# #     Ok(())
# # }
# #
# fn func2() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     let filename = &quot;foo.txt&quot;;
#     do_some_io().map_err(mstrerr!(&quot;Error reading '{}'&quot;, filename))?;
#     Ok(())
# }
# 
# fn func1() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     func2().map_err(mstrerr!(&quot;func1 error&quot;))?;
#     Ok(())
# }
# 
# # fn main() {
# #     if let Err(e) = func1() {
# #         assert_eq!(
# #             format!(&quot;\n{:?}\n&quot;, e), r#&quot;
# # src/lib.rs:20: func1 error
# # Caused by:
# # src/lib.rs:15: Error reading 'foo.txt'
# # Caused by:
# # Kind(NotFound)
# # &quot;#
# #         );
# #     } else {
# #         unreachable!();
# #     }
# # }
# ~~~
# 
# `mstrerr!()` can also be used to map a new `ChainError&lt;T&gt;`, where T was defined with
# `derive_str_cherr!(T)`
# 
# ~~~rust
# # use crate::chainerror::*;
# # use std::error::Error;
# # use std::io;
# # use std::result::Result;
# #
# # fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
# #     Err(io::Error::from(io::ErrorKind::NotFound))?;
# #     Ok(())
# # }
# #
# derive_str_cherr!(Func2Error);
# 
# fn func2() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     let filename = &quot;foo.txt&quot;;
#     do_some_io().map_err(mstrerr!(Func2Error, &quot;Error reading '{}'&quot;, filename))?;
#     Ok(())
# }
# 
# derive_str_cherr!(Func1Error);
# 
# fn func1() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     func2().map_err(mstrerr!(Func1Error, &quot;func1 error&quot;))?;
#     Ok(())
# }
# #
# # fn main() {
# #     if let Err(e) = func1() {
# #         if let Some(f1err) = e.downcast_chain_ref::&lt;Func1Error&gt;() {
# #             assert!(f1err.find_cause::&lt;ChainError&lt;Func2Error&gt;&gt;().is_some());
# #             assert!(f1err.find_chain_cause::&lt;Func2Error&gt;().is_some());
# #         } else {
# #             panic!();
# #         }
# #     } else {
# #         unreachable!();
# #     }
# # }
# ~~~
# **/
# #[macro_export]
# macro_rules! mstrerr {
#     ( $t:ident, $v:expr $(, $more:expr)* ) =&gt; {
#         |e| cherr!(e, $t (format!($v, $( $more , )* )))
#     };
#     ( $t:path, $v:expr $(, $more:expr)* ) =&gt; {
#         |e| cherr!(e, $t (format!($v, $( $more , )* )))
#     };
#     ( $v:expr $(, $more:expr)* ) =&gt; {
#         |e| cherr!(e, format!($v, $( $more , )* ))
#     };
# }
# 
# /** convenience macro to create a &quot;new type&quot; T(String) and implement Display + Debug for T
# 
# ~~~rust
# # use crate::chainerror::*;
# # use std::error::Error;
# # use std::io;
# # use std::result::Result;
# #
# # fn do_some_io() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
# #     Err(io::Error::from(io::ErrorKind::NotFound))?;
# #     Ok(())
# # }
# #
# derive_str_cherr!(Func2Error);
# 
# fn func2() -&gt; ChainResult&lt;(), Func2Error&gt; {
#     let filename = &quot;foo.txt&quot;;
#     do_some_io().map_err(mstrerr!(Func2Error, &quot;Error reading '{}'&quot;, filename))?;
#     Ok(())
# }
# 
# derive_str_cherr!(Func1Error);
# 
# fn func1() -&gt; Result&lt;(), Box&lt;Error&gt;&gt; {
#     func2().map_err(mstrerr!(Func1Error, &quot;func1 error&quot;))?;
#     Ok(())
# }
# #
# # fn main() {
# #     if let Err(e) = func1() {
# #         if let Some(f1err) = e.downcast_chain_ref::&lt;Func1Error&gt;() {
# #             assert!(f1err.find_cause::&lt;ChainError&lt;Func2Error&gt;&gt;().is_some());
# #             assert!(f1err.find_chain_cause::&lt;Func2Error&gt;().is_some());
# #         } else {
# #             panic!();
# #         }
# #     } else {
# #         unreachable!();
# #     }
# # }
# ~~~
# 
# **/
# #[macro_export]
# macro_rules! derive_str_cherr {
#     ($e:ident) =&gt; {
#         struct $e(String);
#         impl ::std::fmt::Display for $e {
#             fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
#                 write!(f, &quot;{}&quot;, self.0)
#             }
#         }
#         impl ::std::fmt::Debug for $e {
#             fn fmt(&amp;self, f: &amp;mut ::std::fmt::Formatter) -&gt; ::std::fmt::Result {
#                 write!(f, &quot;{}({})&quot;, stringify!($e), self.0)
#             }
#         }
#         impl ::std::error::Error for $e {}
#     };
# }
# }
</code></pre></pre>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                        
                            <a rel="prev" href="tutorial10.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>
                        

                        
                            <a rel="next" href="end.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>
                        

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                
                    <a href="tutorial10.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>
                

                
                    <a href="end.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
                
            </nav>

        </div>

        

        

        

        
        <script src="elasticlunr.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="mark.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="searcher.js" type="text/javascript" charset="utf-8"></script>
        

        <script src="clipboard.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="highlight.js" type="text/javascript" charset="utf-8"></script>
        <script src="book.js" type="text/javascript" charset="utf-8"></script>

        <!-- Custom JS scripts -->
        

        

    </body>
</html>
